### PR TITLE
Enable moving back to the root level in the command palette

### LINKED ISSUE #7910: [Functional-Command Palette]- User is not able to move back from the New Tab sub menu as there is no functionality provided to move out of Sub menu.
**User Experience:**
Users who rely on screen reader/Keyboard will not be able to navigate efficiently if there is no option provided to move out of Sub menu

**Test Environment:**
App: Windows Terminal Preview
Feature: Command Palette
Screen Reader: Narrator
OS: Win 10 2004 Build (19041.508)

**Repro Steps:**
1.Open Windows terminal Preview app
2.Open Command Palette using shortcut (Ctrl+shift+p)
3.Search for New tab command in the search field
4.Navigate to New Tab list item having sub menu and activate it .
5.Verify whether there is any option provided to move back.

**Actual**:
User is not able to move back from the sub menu as there is no functionality provided to move out of Sub menu.

Note: Same issue repro for Split pane and Set color scheme controls having Sub menu

**Expected:**
Functionality should be provided to move back from sub menu by using back button or some other way other wise user has to close command palette and open it again to get all commands

### PR DESCRIPTION
**Pull request by Hegunumo**
This commit adds functionality so that users can move back from sub menu
whenever they want. As a result, users no longer have to close command
palette and open it again to get all commands again.

Closes #7910

### CODE REVIEW

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Thanks for throwing this together! I think when I was envisioning this, I was imagining it less as "use backspace to go up a level", but more as an actual button to go up a level. I don't know about other users, but I'm definitely the type of person to just mash the <kbd>Bksp</kbd> key when I want to clear out the input, usually far more than is actually needed. I think I pictured something like this:

image

And clicking on the `<` would take you up the level.

This is the mockup code I had in XAML Studio:

```xaml

            <StackPanel Orientation="Horizontal"
                    Padding="16, 0, 16, 4"
                    Grid.Row="1"> 
                <FontIcon 
                    FontSize="12"
                    FontFamily="Segoe MDL2 Assets" 
                    Glyph="&#xE76b;">
                
                </FontIcon>
                <TextBlock
                    Padding="16, 0, 16, 4"
                    x:Name="_parentCommandText"
                    FontStyle="Italic"
                    Visibility="Visible"
                    Text="New tab with...">
                </TextBlock>
            </StackPanel>

```

**Comment by Hegunumo** [comment]
I hope this is what you meant.

File: src/cascadia/TerminalApp/App.xaml [hunk 1]
```diff
@@ -1,51 +1,59 @@
-<!-- Copyright (c) Microsoft Corporation. All rights reserved. Licensed under
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/App.xaml:1]: Weird that these xaml files keep getting reformatted by your editor - mind reverting them?

File: src/cascadia/TerminalApp/ColorPickupFlyout.xaml [hunk 2]
```diff
@@ -2,106 +2,158 @@
     x:Class="TerminalApp.ColorPickupFlyout"
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/ColorPickupFlyout.xaml:2]: (revert this one too)

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 3]
```diff
@@ -426,6 +426,19 @@ namespace winrt::TerminalApp::implementation
         _dispatchCommand(e.ClickedItem().try_as<Command>());
     }
 
+    void CommandPalette::_moveBackButtonClicked (Windows::Foundation::IInspectable const& /*sender*/,
+                                            Windows::UI::Xaml::RoutedEventArgs const&)
+    {
+        _searchBox().Visibility(Visibility::Collapsed);
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.cpp:432]: I don't think you'll need to hide the search box then make it re-appear, right?

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
I think I was expecting something more like

```xaml

            <StackPanel Orientation="Horizontal"
                    Padding="16, 0, 16, 4"
                    Grid.Row="1"
                    Visibility="{x:Bind ParentCommandName,
                                 Mode=OneWay,
                                 Converter={StaticResource ParentCommandVisibilityConverter}}"> 

                <Button Background="Transparent" x:Name="_parentCommandBackButton" x:Uid="ParentCommandBackButton"> 
                    <FontIcon 
                        FontSize="12"
                        FontFamily="Segoe MDL2 Assets" 
                        Glyph="&#xE76b;">
                    </FontIcon>
                </Button>
                
                <TextBlock
                    Padding="16, 0, 16, 4"
                    x:Name="_parentCommandText"
                    FontStyle="Italic"
                    Grid.Row="1"
                    Text="{x:Bind ParentCommandName, Mode=OneWay}">
                </TextBlock>
            </StackPanel>
```

That way, command palette shouldn't need to futz with editing the text of the parent command to prepend the `<`, it should _just work_. GrantedI've only tested this in XAML Studio, so I'm not positive this will work right.

Also - make sure to add 
```xml

  <data name="ParentCommandBackButton.[using:Windows.UI.Xaml.Automation]AutomationProperties.Name" xml:space="preserve">
    <value>Back</value>
  </data>
  <data name="ParentCommandBackButton.[using:Windows.UI.Xaml.Controls]ToolTipService.ToolTip" xml:space="preserve">
    <value>Back</value>
  </data>
```

to the `Resources.resw` file, so the button will have a tooltip and appear in UIA

File: src/cascadia/TerminalApp/ColorPickupFlyout.xaml (hunk 2 above)
Comment (Hegunumo) [reply src/cascadia/TerminalApp/ColorPickupFlyout.xaml]: I think it's something to do with clang format, but I don't know how to fix it.

**Comment by Hegunumo** [comment]
“Resources.resw file, so the button will have a tooltip and appear in UIA”
I could not fully understand what this mean even though I added to the file.
What is tooltip and UIA?

**Comment by zadjii-msft** [comment]
> What is tooltip and UIA?

A tooltip is the little helper text that appears over an item when you hover it with the mouse. For example:
image
That "Close" text is the _tooltip_ for the close button.

UIA in this context refers to "UI Automation", which is a set of APIs that allow certain accessibility tools to navigate UIs. This enables screen readers to be able to list the name of the control, so users who are visually impaired can know what the control is

**Review by zadjii-msft — APPROVED** [review]
Looks good to me, thanks!

**Comment by zadjii-msft** [comment]
(make sure to run `tools\runformat.cmd` from the root of the repo to make sure the code is formatted correctly)

**Review by miniksa — APPROVED** [review]
Looks good to me.

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 4]
```diff
@@ -426,6 +426,25 @@ namespace winrt::TerminalApp::implementation
...
+    // Method Description:
+   // This event is called when the user clicks on an ChevronLeft button right
+   // next to the ParentCommandName (e.g. New Tab...) above the subcommands list.
+   // It'll go up a level when the users click the button.
+   // Arguments:
+   // - sender: the button that got clicked
+   // Return Value:
+   // - <none>
+    void CommandPalette::_moveBackButtonClicked(Windows::Foundation::IInspectable const& /*sender*/,
+        Windows::UI::Xaml::RoutedEventArgs const&)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.cpp:438]:
```suggestion
    void CommandPalette::_moveBackButtonClicked(Windows::Foundation::IInspectable const& /*sender*/,
                                                Windows::UI::Xaml::RoutedEventArgs const&)
```

**Comment by zadjii-msft** [comment]
@Hegunumo please make sure to run `tools\runformat.cmd` from the root of the repo to make sure the code is formatted correctly ☺️

**Comment by Hegunumo** [comment]
Run it right before submit PR or right after forked the repository ?
Are these 3 failings due to code formatting ?
Do 8 checks always have to be successful in order to get merged ?
By the way, thank you for you help.

**Comment by zadjii-msft** [comment]
You can just run the script locally, then commit & push the changes. The script will auto-format the code for you. This will resolve the "Code Health Scripts Proper Code Formatting Check" check, which will also in turn resolve the "Terminal CI" check. 

We're also just gonna get rid of the "Lint Code Base" step, because it is incredibly flaky. So don't worry about that.

**Review by DHowett — APPROVED** [review]
I love it. Thank you!

**Comment by DHowett** [comment]
Eventually we can make it _pop_ the action stack instead of going up to the root, but this is way better than it was :smile:

**Pull request merged**